packages/backend-core/src/context/index.ts

```diff
@@ -0,0 +1,247 @@
+import env from "../environment"
+import { SEPARATOR, DocumentTypes } from "../db/constants"
+import cls from "./FunctionContext"
+import { dangerousGetDB, closeDB } from "../db"
+import { baseGlobalDBName } from "../tenancy/utils"
+import { IdentityContext } from "@budibase/types"
+import { DEFAULT_TENANT_ID as _DEFAULT_TENANT_ID } from "../constants"
+import { ContextKeys } from "./constants"
+import {
+  updateUsing,
+  closeWithUsing,
+  setAppTenantId,
+  setIdentity,
+  closeAppDBs,
+  getContextDB,
+} from "./utils"
+
+export const DEFAULT_TENANT_ID = _DEFAULT_TENANT_ID
+
+// some test cases call functions directly, need to
+// store an app ID to pretend there is a context
+let TEST_APP_ID: string | null = null
+
+export const closeTenancy = async () => {
+  let db
+  try {
+    if (env.USE_COUCH) {
+      db = getGlobalDB()
+    }
+  } catch (err) {
+    // no DB found - skip closing
+    return
+  }
+  await closeDB(db)
+  // clear from context now that database is closed/task is finished
+  cls.setOnContext(ContextKeys.TENANT_ID, null)
+  cls.setOnContext(ContextKeys.GLOBAL_DB, null)
+}
+
+// export const isDefaultTenant = () => {
+//   return getTenantId() === DEFAULT_TENANT_ID
+// }
+
+export const isMultiTenant = () => {
+  return env.MULTI_TENANCY
+}
+
+/**
+ * Given an app ID this will attempt to retrieve the tenant ID from it.
+ * @return {null|string} The tenant ID found within the app ID.
+ */
+export const getTenantIDFromAppID = (appId: string) => {
+  if (!appId) {
+    return null
+  }
+  const split = appId.split(SEPARATOR)
+  const hasDev = split[1] === DocumentTypes.DEV
+  if ((hasDev && split.length === 3) || (!hasDev && split.length === 2)) {
+    return null
+  }
+  if (hasDev) {
+    return split[2]
+  } else {
+    return split[1]
+  }
+}
+
+// used for automations, API endpoints should always be in context already
+export const doInTenant = (tenantId: string | null, task: any) => {
+  // the internal function is so that we can re-use an existing
+  // context - don't want to close DB on a parent context
+  async function internal(opts = { existing: false }) {
+    // set the tenant id + global db if this is a new context
+    if (!opts.existing) {
+      updateTenantId(tenantId)
+    }
+
+    try {
+      // invoke the task
+      return await task()
+    } finally {
+      await closeWithUsing(ContextKeys.TENANCY_IN_USE, () => {
+        return closeTenancy()
```

Comment:
I think we should await here for closeTenancy, just in case it takes longer for any reason. does it make sense?

Reply:
here this isn't an async function within, its just straight returning the `closeTenancy` function, so the promise will simply be returned for the parent (wrapper) to handle.

Reply:
Oh I see! thanks for the details :)